components/net/http_cache.rs

```diff
@@ -562,6 +580,126 @@ impl HttpCache {
+    pub fn invalidate(&mut self, request: &Request, response: &Response) {
+        // TODO(eijebong): Once headers support typed_get, update this to use them
+        if let Some(Ok(location)) = response
+            .headers
+            .get(header::LOCATION)
+            .map(HeaderValue::to_str)
+        {
+            if let Ok(url) = request.current_url().join(location) {
+                self.invalidate_for_url(&url);
+            }
+        }
+        if let Some(Ok(ref content_location)) = response
+            .headers
+            .get(header::CONTENT_LOCATION)
+            .map(HeaderValue::to_str)
+        {
+            if let Ok(url) = request.current_url().join(&content_location) {
+                self.invalidate_for_url(&url);
+            }
+        }
+        self.invalidate_for_url(&request.url());
+    }
+}
+
+fn check_vary_headers(request: &Request, cached_resource: &CachedResource) -> bool {
+    let mut can_be_constructed = true;
+    let cached_headers = cached_resource.data.metadata.headers.lock().unwrap();
+    let original_request_headers = cached_resource.request_headers.lock().unwrap();
+    if let Some(vary_value) = cached_headers.typed_get::<Vary>() {
+        if vary_value.is_any() {
+            can_be_constructed = false
+        } else {
+            // For every header name found in the Vary header of the stored response.
+            // Calculating Secondary Keys with Vary <https://tools.ietf.org/html/rfc7234#section-4.1>
+            for vary_val in vary_value.iter_strs() {
+                match request.headers.get(vary_val) {
+                    Some(header_data) => {
+                        // If the header is present in the request.
+                        if let Some(original_header_data) = original_request_headers.get(vary_val) {
+                            // Check that the value of the nominated header field,
+                            // in the original request, matches the value in the current request.
+                            if original_header_data != header_data {
+                                can_be_constructed = false;
+                                break;
+                            }
+                        }
+                    },
+                    None => {
+                        // If a header field is absent from a request,
+                        // it can only match a stored response if those headers,
+                        // were also absent in the original request.
+                        can_be_constructed = original_request_headers.get(vary_val).is_none();
+                    },
+                }
+                if !can_be_constructed {
+                    break;
+                }
+            }
+        }
+    }
+    can_be_constructed
+}
+
+#[derive(Clone)]
+/// A cache entry.
+pub struct HttpCacheEntry {
+    /// Resources corresponding to the entry.
+    pub resources: Arc<RwLock<Vec<Arc<RwLock<CachedResource>>>>>,
```

Comment:
note: wondering if we should not try using a `HashSet` instead of a `Vec`, since the "ordered" iteration over resources would seem to, if anything, reduce opportunities to parallelism.

Reply:
Let's forget about this: too complicate to make `CachedResource` something that can be put in a set.